Refactor GSI response tracker

jdom opened the pull request:
Simplify GlobalSingleInstanceReponseTracker abstraction.
All the response tracking logic was encapsulated into `GetOutcome` and `GetOutcomeAsync`, with a much clearer responsibility defined by the interface.
This greatly simplifies caller code to avoid using a custom `TaskCompletionSource` and remembering to call into `CheckIfDone()` after mutating shared state (the array that was passed into its constructor), where the responsibilities where not clear. Also avoids using tasks that in theory should be completed (but not enforced by the previous abstraction) when the method returns, when passing the complete batch of responses (instead of an array of responses that would be filled in as responses come back).

jdom committed:
Compose TCS instead of inheriting

jdom committed:
Refactor to use promises

jdom committed:
Refactor GSI response tracker

jdom commented:
This would address my comments here: https://github.com/dotnet/orleans/pull/1800#discussion_r71052816 (and others related to this tracker class).

I kept as 3 commits so that changes are less drastic, but feel free to squash into 1 when merging it to your branch (via the Github UI)

sebastianburckhardt commented on src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceResponseTracker.cs:
```diff
@@ -50,29 +51,31 @@ public override string ToString()
-                if (responses.All(res => res != null && res.ResponseStatus == ActivationResponseStatus.Pass))
+                // store incomplete promises at this time (as they might be completed by the time the method finishes
+                var incompletePromises = responses.Where(t => !t.IsCompleted).ToArray();
+                if (incompletePromises.Length == 0 && responses.All(res => res.IsCompleted && res.Result.ResponseStatus == ActivationResponseStatus.Pass))
```
I think checking the condition `res.IsCompleted` is redundant (because we already check `incompletePromises.Length ==0`)

sebastianburckhardt commented on src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceResponseTracker.cs:
```diff
@@ -21,15 +21,16 @@ public enum Outcome {
+        public GlobalSingleInstanceResponseTracker(Task<RemoteClusterActivationResponse>[] responses, GrainId grain, Logger logger)
         {
             this.responses = responses;
+            Debug.Assert(this.responses.All(t => t != null));
```
I think we generally don't use Debug.Assert in the codebase.

jdom commented:
Those comments were in the intermediate diff, but not on the final one. 
For example, the Debug.Assert code was there before I rebased (I started this work on Friday)

sebastianburckhardt commented:
Yes, I noticed right after I commented... I am still in the process of figuring out how to best review a PR. Pulling the branch off your fork worked, so I am now looking at a diff locally on my machine (the git web diff interface seems not ideal).

So far everything looks great, I will have a closer look, run the tests, and then merge. Thanks for improving the response tracker code.

jdom commented:
OK. I only ran de GeoCluster tests and they all pass. 
A nice alternative to review medium and large PRs is to use CodeFlow, which now natively supports github pull requests

sebastianburckhardt commented on src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceResponseTracker.cs:
```diff
@@ -1,107 +1,152 @@
         }
 
-        // for tracing, display outcome
-        public override string ToString()
```
Instead of removing this `ToString()` method altogether, perhaps it should be moved to `GlobalSingleInstanceResponseOutcome`. The information it provides in the trace has been helpful to me in the past.

sebastianburckhardt commented on src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceRegistrar.cs:
```diff
@@ -95,29 +95,27 @@ public void Delete(GrainId gid)
 
                 if (logger.IsVerbose)
-                    logger.Verbose("GSIP:End {0} Round={1} Outcome={2}", address.Grain.ToString(), numRetries - retries, responses);
+                    logger.Verbose("GSIP:End {0} Round={1} Outcome={2}", address.Grain.ToString(), numRetries - retries, outcome.State);
```
Did you mean to make the tracing message less informative? It no longer prints the activation address if a remote owner is found. Before, this called a custom ToString() method on the response tracker that included the activation address.

jdom replied:
no, I did not, I thought the ToString was mostly to check whether the tracking completed or not (and at this point after the await it was always completed), and I missed that. So I'll re-add this information in the `GlobalSingleInstanceResponseOutcome` class as you suggested in your other comment

jdom replied:
addressed the 3 comments, please refresh

sebastianburckhardt commented on src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceActivationMaintainer.cs:
```diff
@@ -250,40 +250,35 @@ private Task RunBatchedDemotion(List<Tuple<GrainId, KeyValuePair<ActivationId, I
+                var outcome = outcomeDetails.State;
 
                 if (logger.IsVerbose2)
                     logger.Verbose2("GSIP:M {0} Result={1}", address.Grain.ToString(), outcome.ToString());
```
if we add the `ToString()` that prints a remotely owned activation address we can also use `outcomeDetails.ToString()` here.

jdom committed:
Trace extended information about GSI owner response

sebastianburckhardt commented on src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceResponseTracker.cs:
```diff
@@ -1,107 +1,157 @@
+
+        public override string ToString()
+        {
+            return $"[{this.State} {this.RemoteOwnerAddress.Address}]";
```
Just to make sure: could this throw a null pointer exception if RemoteOwnerAddress.Address is null?
(previously there was a check)

sebastianburckhardt replied:
I think string interpolation inserts empty strings for null values, so this should be o.k.

jdom replied:
no, passing a null to a string.format (or in this case using string interpolation which is similar) is legal.

jdom commented:
Hi, thanks for merging. BTW, I didn't realize you merged, but I added an additional commit fixing code comments? Do you want me to create a separate PR (I can easily do this) or do you want to grab the changes directly?
https://github.com/jdom/orleans/commit/254014be643f0149c2c8adf540d3788a06dc8afd

EDIT: updated link to additional commit

sebastianburckhardt commented:
I can grab them directly, that’s easier.

Thanks,
Sebastian

From: Julian Dominguez [mailto:notifications@github.com]
Sent: Wednesday, September 7, 2016 10:39 AM
To: sebastianburckhardt/orleans orleans@noreply.github.com
Cc: Sebastian Burckhardt sburckha@microsoft.com; State change state_change@noreply.github.com
Subject: Re: [sebastianburckhardt/orleans] Refactor GSI response tracker (#5)

Hi, thanks for merging. BTW, I didn't realize you merged, but I added an additional commit fixing code comments? Do you want me to create a separate PR or do you want to grab the changes directly?
ea1b397https://github.com/sebastianburckhardt/orleans/pull/5/commits/ea1b397ad5094004e6b449361d0b283f067f86b2

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHubhttps://github.com/sebastianburckhardt/orleans/pull/5#issuecomment-245359039, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AGHlkJf3u8_05QPfhZKiJp7KJAt_73xRks5qnvaagaJpZM4J2Xai.
